Add snippet feature

gdolle opened the pull request:
Add snippet code feature similarly to [doxygen ability](https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdsnippet).
See also the [doxygen document guide](https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html) for comment style.

Example:

We consider the following C++ code,

``` cpp
// test.cpp
#include <iostream>
/** [marker0] */
int main()
{
   //! [marker1]
   int a=0;
   //! [marker1]
   /// [marker2]
   int b=1;
   int c=2;
   /// [marker2]
   return 0;
}
/** [marker0] */
```

then both slice can be imported via 
- `[import:marker1](path/to/test.cpp)`

``` cpp
   int a=0;
```
- `[import:marker2](path/to/test.cpp)`

``` cpp
   int b=1;
   int c=2;
```
- `[include:marker0](path/to/test.cpp)` (updated from @prudhomm comment)

``` cpp
int main()
{
   int a=0;
   int b=1;
   int c=2;
   return 0;
}
```

and does not depend on lines number in that case.

NB: C/C++, latex, python supported for now but could be improved.

gdolle committed:
Add snippet feature

prudhomm commented:
very nice !

azu commented on src/parser.js:
```diff
@@ -53,12 +53,34 @@ export function getSliceRange(label) {
+
+ [import:markname, hello-world.js](../src/hello-world.js)
+ */
+export function getMarkerName(label) {
```
It should be moved to `marker.js`(for testing).

And split it into `getMarkerName(label)` and `hasMarker(label)` function.

azu commented on src/parser.js:
```diff
@@ -53,12 +53,34 @@ export function getSliceRange(label) {
     const content = slicedCode.trim();
-    return generateEmbedCode(lang, fileName, originalPath, content);
+    const content2 = markersSliceCode( code, marker );
+    if(marker!='') {
```
add `hasMarker(label)` and use it.

azu commented on src/parser.js:
```diff
@@ -84,6 +106,32 @@ function sliceCode(code, start, end) {
     return splitted.slice(start - 1, end).join('\n');
 }
 
+export function markersSliceCode( code, markername )
```
This function is complex :-)
I think need to test :green_heart: 

e.g.) add marker-test.js and test it like [gitbook-plugin-include-codeblock/language-detection-test.js at master · azu/gitbook-plugin-include-codeblock](https://github.com/azu/gitbook-plugin-include-codeblock/blob/master/test/language-detection-test.js)

azu replied:
Basic example.

``` js
const assert = require("power-assert");
import {getMarker, hasMarker, markersSliceCode} from "../src/marker";
describe("marker", function () {
    // we need to good condition and bad condition.
    // e.g.) when has not marker
    context("when has [marker1]", function(){
        it("should slice code between [marker1] and [marker1]", function(){
            const code = `
// test.cpp source code
int main()
{
    int a;
    // [marker1]
    int b;
    // [marker1]
    int c;
}
            `;
            const markerName = "marker1";
            const result = markersSliceCode(code, markerName);
            const expectedSlicedCode = `    int b;`// right?
            assert.equal(result, expectedSlicedCode);
        });
    });
});
```

azu commented on src/parser.js:
```diff
@@ -84,6 +106,32 @@ function sliceCode(code, start, end) {
+    }
+    else {
+        // various language comment
+        var commentOpen="(\/+\/+|#|%|\/\\*)";
```
use `const` instead of `var`.
(This code is transpiled by [Babel](https://babeljs.io/))

azu commented on src/parser.js:
```diff
@@ -84,6 +106,32 @@ function sliceCode(code, start, end) {
+        }
+        else {
+            console.warn('marker `'+markername+'` not found');
+            return [];
```
It should be `throw new Error('marker`'+markername+'`not found')`.

Tips: it is smart that use [Template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals).

``` js
throw new Error(`marker:${markername} not found`);
```

OR 

return warning message(it will be shown in html file)

``` js
console.warn('marker `'+markername+'` not found');
return 'marker `'+markername+'` not found'
```

azu commented:
@gdolle Thanks for nice feature!

Summary after I've reviewed.
- Split up marker feature to `marker.js` like [language-detection.js](https://github.com/azu/gitbook-plugin-include-codeblock/blob/master/src/language-detection.js).
  - (Of course, Slice feature also should be separated... After merging this pull request, I work on it)
- Need tests.

azu commented on src/parser.js:
```diff
@@ -84,6 +106,32 @@ function sliceCode(code, start, end) {
+        var balise="\\["+markername+"\\]";
+        var pattern=commentOpen + ".*\\s*" + balise + ".*\\s*" + commentClose;
+        var regstr=pattern+"([\\s\\S]*)"+pattern;
+        var reg = new RegExp(regstr);
```
:memo: https://github.com/jonschlinkert/parse-comments parse code comments from any lang.
It will be improved in the future.

prudhomm commented:
@gdolle @azu it would be nice to have an option to actually remove the `// [marker1]` lines from being included so that we have a nice clean code example.

consider a code that has all the markers, we go through all of them to explain the code (we don't show the markers). and then at the end we want to put the entire code. it would not be very nice to have all these `// [marker]` and there should be a flag to automatically remove them (enabled by default I would say).

gdolle committed:
Split marker implementation

Implement Reviewed comments.

- Create hasMarker in separate marker.js file
- Move getMarkerName in marker.js
- Change to const variables

gdolle committed:
Fix require

gdolle committed:
Refactor code

- Update and clean code in src/marker.js
- Add removeMarkers function which remove inner markers
  (@prudhomm request)
- Add some code comments.
- Change comment style to better respect doxygen documenting style.
  The language comment keyword must be followed by either !,* or /
  as described here
    https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html.
  and for snippet balise here:
    https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdsnippet
  See the bellow example.

Example:

```cpp
// C++ example

int main(){
/** [marker0] */
//! [marker1]
    int a;
/// [marker1]
    int b;
/// [marker2]
    // Simple C++ comment.
    int c;
//! [marker2]
/** [marker0] */
return 0;
}

```
will result to

```
    int a;
    int b;
    // Simple C++ comment.
    int c;
```

gdolle committed:
Add marker unit tests

gdolle committed:
Update espower-babel

- espower-babel is deprecated, see
  https://github.com/power-assert-js/espower-babel
- Update babel to 6.x.x

gdolle committed:
Update README

gdolle commented:
@azu @prudhomm I updated following the review comments, meaning:
- splitting the code in a separate file src/marker.js
- add several unit tests in test/marker-test.js
- fix const variable
- add hasMarker() function
- add removeMarker() function which remove inner markers following the suggestion of @prudhomm.

Note that I updated espower-babel which is [deprecated](https://github.com/power-assert-js/espower-babel#warning-deprecated-warning) now. (Test pass see travis. ;) )

azu commented on test/marker-test.js:
```diff
@@ -0,0 +1,127 @@
+const expectedMarker6 = `    int h;`;
+
+
+describe("marker", function () {
```
:green_heart: Good!

azu commented:
@gdolle Thanks for great work!
I've merged it and publish [1.8.0](https://github.com/azu/gitbook-plugin-include-codeblock/releases/tag/1.8.0).

azu commented:
After merging, I found a breaking backward compatibily.
I've opened pull request to fix that issue #13

azu commented:
I've fixed regression issue #13  and publish [1.8.1](https://github.com/azu/gitbook-plugin-include-codeblock/releases/tag/1.8.1).

Thanks again @gdolle @prudhomm
